design/backup-resources-order.md

```diff
@@ -0,0 +1,39 @@
+## Backup Resources Order
+This document proposes a solution that allows user to specify a backup order for resources of specific resource type.
+
+## Background
+During backup process, user may need to back up resources of specific type in some specific order to ensure the resources were backup properly because these resources are related and ordering might be required to preserve the consistency for the apps to recover itself �from the backup image 
+(Ex: master-slave database pods in a cluster).
+
+## Goals
+- Enable user to specify an order of back up resources belong to specific resource type
+
+## Alternatives Considered
+- Use a plugin to backup an resources and all the sub resources.  For example use a plugin for StatefulSet and backup pods belong to the StatefulSet in specific order.  This plugin solution is not generic and requires plugin for each resource type.
+
+## High-Level Design
+User will specify a map of resource type to list resource names (separate by semicolons).  Each name will be in the format "namespaceName/resourceName" to enable ordering accross namespaces.  Based on this map, the resources of each resource type will be sorted by the order specified in the list of resources.  If a resource instance belong to that specific type but its name is not in the order list, then it will be put behind other resources that are in the list.

```

Comment:
How could cluster-scoped resources be handled? Unfortunately, I don't have a concrete example here, other than some theoretical CRs that may depend on each other.

This may or may not be a concern, but I do want to ask. If the theoretical CRDs worked like PVs, there would be a namespaced resource that is similar to a PVC pointing to them that could be used to order instead.

Reply:
In my usage, we only work within a context of a namespace.  However, this design should work across namespace or in the cluster scope because the resourceName is specified as "namespace/resourceName".  If the resources have cluster scopes (for example PersistentVolumes), then we simply use pvName without namespace prefix.  The implementation is the same because we look up the resource from the resource map based on the resourceName.